docs/notes/shutdown.txt

```diff
@@ -0,0 +1,68 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+
+
+qdrouterd shutdown procedure
+============================
+
+
+How does the router shut down when interrupted with ctrl-C ?
```

Comment:
This probably needs some kind of overview section: what are the things that actually need to get shut down? In what order?

How do you shut down a Proton proactor loop? Why can't `main.c::signal_handler()` simply call `pn_proactor_interrupt()` in a loop as many times as there are threads? Why doesn't Proactor have a way to kill all threads at once?

How do you shut down memory pool? How do you shut down Python interpreter. How do you shut down libwebsockets and libnghttp2 listeners?

Can you do graceful shutdown, quiescing? Router will stop receiving new connections and will stop when last connection disconnects? Afaik no, the router shutdown will forcibly kill all connections.

Reply:
I bet it's easy to see bugs with those big glasses.

Reply:
I meant you should have something high-level about what needs to happen so that the router shutdowns.

* stop proactor,
* close connections,
* stop adaptors (yeah, you have nothing about adaptors here, and they do have their own lifecycle),
* and so on

Maybe also _why_ is it happening. But that can be another docs, further on the route towards fixing, yeah.

> In this document, I would rather only describe what is happening.

+1, I am all for gradual staged approach